Add support for configuring Dask distributed (#2049)
ESMValGroup/ESMValCore: doc/quickstart/configure.rst

parallel.
In order to figure out what needs to be computed when, Dask makes use of a
'`scheduler <https://docs.dask.org/en/stable/scheduling.html>`_'.
The default scheduler in Dask is rather basic, so it can only run on a single
computer and it may not always find the optimal task scheduling solution,
resulting in excessive memory use when using e.g. the
:func:`esmvalcore.preprocessor.multi_model_statistics` preprocessor function.
Therefore it is recommended that you take a moment to configure the
`Dask distributed <https://distributed.dask.org>`_ scheduler.
A Dask scheduler and the 'workers' running the actual computations, are
collectively called a 'Dask cluster'.

In ESMValCore, the Dask cluster can configured by creating a file called
``~/.esmvaltool/dask.yml``, where ``~`` is short for your home directory.
In this file, under the ``client`` keyword, the arguments to
:obj:`distributed.Client` can be provided.
Under the ``cluster`` keyword, the type of cluster (e.g.
:obj:`distributed.LocalCluster`), as well as any arguments required to start
the cluster can be provided.

Below are some example configurations:

> Create a Dask distruted cluster on the computer running ESMValCore using

Reviewer: ```suggestion
Create a Dask distributed cluster on the computer running ESMValCore using
```

Reviewer 2: "distruted" is a perfectly fine word, look it up in Oxford Dictionary :rofl:

all available resources:

.. code:: yaml

  cluster:
    type: distributed.LocalCluster

this should work well for most personal computers.

.. note::

   Note that, if running this configuration on a shared node of an HPC cluster,
   Dask will try and use as many resources it can find available, and this may
   lead to overcrowding the node by a single user (you)!

> Create a Dask distruted cluster on the computer running ESMValCore, with

Reviewer: ```suggestion
Create a Dask distributed cluster on the computer running ESMValCore, with
```

4 workers with two 2 GiB of memory each (8 GiB in total):

.. code:: yaml

  cluster:
    type: distributed.LocalCluster
    n_workers: 4
    memory_limit: 2 GiB

Create a Dask distributed cluster on the
`Levante <https://docs.dkrz.de/doc/levante/running-jobs/index.html>`_
supercomputer using the `Dask-Jobqueue <https://jobqueue.dask.org/en/latest/>`_

> package:

Reviewer 3: It would be nice to mention that this needs to be installed by the user (e.g., `mamba install dask-jobqueue`) because it's not part of our environment.

Author: I like @valeriupredoi's suggestion of just adding it to the dependencies. It doesn't have any dependencies that we do not already have and it's a very small Python package.

Reviewer 3: Sounds good, that's even better!

Author: Done in [25dc5ce](https://github.com/ESMValGroup/ESMValCore/pull/2049/commits/25dc5ce6f7e39e8a9b02f0777bcdcd70b4f2328e)

.. code:: yaml

  cluster:
    type: dask_jobqueue.SLURMCluster
    queue: interactive
    account: bk1088
    cores: 8

>     memory: 16GiB

Reviewer 4: Do you know if there is any way to avoid having to specify the memory requirements? 

I am not allowed to submit jobs with memory requirements in our infrastructure, because they are set by default depending on the type of node you are running with. So my job submissions fail because of that. 

However, the configuration seems to require `memory` to be specified. I am just wondering if there is any way to set it to null or something.

*Personal computer*

Create a Dask distributed cluster on the computer running ESMValCore using
all available resources:

.. code:: yaml

  cluster:
    type: distributed.LocalCluster

this should work well for most personal computers.

.. note::

   Note that, if running this configuration on a shared node of an HPC cluster,
   Dask will try and use as many resources it can find available, and this may
   lead to overcrowding the node by a single user (you)!

*Shared computer*

Create a Dask distributed cluster on the computer running ESMValCore, with
2 workers with 4 threads/4 GiB of memory each (8 GiB in total):

.. code:: yaml

  cluster:
    type: distributed.LocalCluster
    n_workers: 2
    threads_per_worker: 4
    memory_limit: 4 GiB

this should work well for shared computers.

*Computer cluster*

Create a Dask distributed cluster on the
`Levante <https://docs.dkrz.de/doc/levante/running-jobs/index.html>`_
supercomputer using the `Dask-Jobqueue <https://jobqueue.dask.org/en/latest/>`_
package:

.. code:: yaml

  cluster:
    type: dask_jobqueue.SLURMCluster
    queue: shared
    account: bk1088
    cores: 8
    memory: 7680MiB
    processes: 2
    interface: ib0

>     local_directory: "/work/bd0854/b381141/dask-tmp"

Reviewer 5: The Dask config should accept environment variables, maybe useful to show this for the following?
```suggestion
    local_directory: "$SCRATCH"
```

Author: Thanks for reviewing @fnattino! I will add a note to the text, but not change the configuration example here because this [environmental variable is not provided on Levante](https://docs.dkrz.de/doc/levante/file-systems.html) so then the example wouldn't work.

   debugging, etc. You can even provide any config user value as a run flag
   ``--argument_name argument_value``

.. _config-dask:

Dask distributed configuration
==============================

The :ref:`preprocessor functions <preprocessor_functions>` and many of the
:ref:`Python diagnostics in ESMValTool <esmvaltool:recipes>` make use of the
:ref:`Iris <iris:iris_docs>` library to work with the data.
In Iris, data can be either :ref:`real or lazy <iris:real_and_lazy_data>`.
Lazy data is represented by `dask arrays <https://docs.dask.org/en/stable/array.html>`_.
Dask arrays consist of many small
`numpy arrays <https://numpy.org/doc/stable/user/absolute_beginners.html#what-is-an-array>`_
(called chunks) and if possible, computations are run on those small arrays in
parallel.
In order to figure out what needs to be computed when, Dask makes use of a
'`scheduler <https://docs.dask.org/en/stable/scheduling.html>`_'.
The default scheduler in Dask is rather basic, so it can only run on a single
computer and it may not always find the optimal task scheduling solution,

> resulting in excessive memory use when using e.g. the

Reviewer 2: ```suggestion
resulting in excessive memory use when running an already memory-intensive task like e.g. the
```

Author: The problem isn't so much that it's memory-intensive, but that the task graph becomes too complicated for the built-in scheduler.

:func:`esmvalcore.preprocessor.multi_model_statistics` preprocessor function.
Therefore it is recommended that you take a moment to configure the
`Dask distributed <https://distributed.dask.org>`_ scheduler.
A Dask scheduler and the 'workers' running the actual computations, are

> together called a Dask 'cluster'.

Reviewer 2: ```suggestion
collectively called a 'Dask cluster'.
```

In ESMValCore, the Dask cluster can configured by creating a file called
``~/.esmvaltool/dask.yml``, where ``~`` is short for your home directory.
In this file, under the ``client`` keyword, the arguments to
:obj:`distributed.Client` can be provided.
Under the ``cluster`` keyword, the type of cluster (e.g.
:obj:`distributed.LocalCluster`), as well as any arguments required to start
the cluster can be provided.

Below are some example configurations:

Create a Dask distruted cluster on the computer running ESMValCore using
all available resources:

.. code:: yaml

  cluster:
    type: distributed.LocalCluster

> this should work well for most personal computers.

Reviewer 2: ```suggestion
this should work well for most personal computers. Note that, if running this configuration on a shared node of an HPC cluster, Dask will try and use as many resources it can find available, and this may lead to overcrowding the node by a single user (you)!
```

Reviewer 2: or, maybe encase this in a .. note::
